### ISSUE TITLE
OSX Mountain Lion MIDI IN issues

### ISSUE THREAD

**Issue by charredUtensil** [samphonic/MoppyClassic#49]
I've managed to put together a humble little setup of 4 drives, which play very nicely under Windows and OSX with MIDI files. This much is flawless. However, as I do not own Logic, my favorite music editing software by far is Apple's GarageBand. It has only one major flaw - it does not let you export MIDI. However, I've found a solution:

1) Set up the Apple IAC driver with a few virtual midi cables
2) Use the MidiO component as an instrument in GB to play the tracks to the IAC driver
3) Start Moppy-Advanced, set to play MIDI IN from the IAC driver

Unfortunately, attempting to use the playable keyboard  in GB will cause Moppy to play a single note that will never stop (until Moppy is disconnected), or a single note for about half a second before refusing to play any further notes.

I have tried using Moppy with the included MIDI synthesizers instead of connecting the floppy drives, but this does not help. The most bizarre part of this was that it _was working_ for about 20 minutes, then suddenly when I started playing the song, each drive started playing the first note, stopped playing the first note, and, upon reaching the second note, played the first note again forever until Moppy was disconnected, and now Moppy refuses to play any notes again.

Any ideas? Suggestions? I know this is a series of terrible hacks that would just be alleviated if Apple could put such an important feature in their products, but I would expect nothing less from them. :) My suspicion is that something still thinks it is playing a note, even though I restarted everything twice.

**Comment by samphonic** [comment]
Sorry it's taken me so long to get back to you on this.

My best guess off-hand is that Apple's driver is sending "technically-correct" MIDI messages, but in a way that confuses the code I wrote.  For example, I found that in some cases, rather than sending a "note on" event, followed by a "note off" event, some synths will send a "note on at 127 (max) velocity" followed by a "note on at 0 velocity".  Which, effectively makes the note "stop" playing, but without using the "note off" command.

I have that particular case handled, now, but maybe Apple's doing some weird stuff.  If you put a breakpoint on line 56 of MoppyPlayerOutput.java and use "Debug" instead of "Run" in NetBeans, you should be able to examine the MidiMessage objects that are being received by the Java software and being sent to the Arduino.  If you play a single note on the software keyboard, you should be able to step through the messages one at a time to see if anything weird comes through.

I know there might be a bit of a learning curve here as far as debugging Java software in NetBeans if you haven't before, but it's not as scary as it sounds, and if you run into any specific questions, I'll try to help as best I can from here

**Comment by samphonic** [comment]
Oh!  I forgot to add this helpful link:
http://www.midi.org/techspecs/midimessages.php

Which will help give you a better idea of what the messages mean.

**Comment by charredUtensil** [comment]
So I actually ended up rewriting a command line version of Moppy's controller software in Ruby, as:
1. Let's face it. There's a lot of code in Moppy right now and I was trying to see if I could get it done in only a few lines of code.
2. I'm very familiar with Ruby at the moment and haven't used Java in a good long while
3. The Ruby libraries I found seem to work very well on OSX and Linux, and installed perfectly fine with Bundler (a package manager for Ruby 'gems'), while I had one hell of a time trying to install the serial controller stuff on OSX

At the moment, it only works when bound to live MIDI, hasn't been successfully tested, and will only work on POSIX systems, as I hardcoded it to check /dev for serial modems instead of using anything in the library, so I'm not quite ready to push my changes upstream. I plan to do so once it either reads MIDI files or I can get the stream feature tested. In doing this, I did discover the problem:

MidiO, for whatever reason, is _spamming_ MIDI control messages. I attempted to play a simple sequence, C4 D4 E4. The output to the pipeline, at the end of the sequence, was a random assortment of the following messages, sent in a random order _every few milliseconds_:

```
PROGRAM CHANGE
NOTE ON C4 127
NOTE OFF C4
NOTE ON D4 127
VELOCITY CHANGE C4 = 0
RESERVED
NOTE OFF E4
NOTE OFF D4
NOTE ON E4 127
RESERVED
```

Therefore, the problem is in MidiO and not Moppy. The drawback to this is that MidiO seems to be the only way to get MIDI out of GarageBand, hasn't been updated since 2004, is not open sourced, and has no official repository or issue tracking outside emailing the author. So, your guess was pretty accurate, but the driver isn't sending technically correct MIDI messages. It's sending something about one tier above line noise.

I can close this issue or leave it open for anyone else trying to use live MIDI streaming with Garageband or on OSX.

**Comment by charredUtensil** [comment]
Closing the issue because I just attempted this again after not touching it for a week (the wonders that works sometimes!) I tried everything I could, including reinstalling the midiO plugin. I used a different source to download the .component file and this source came with a more detailed README than the original source I found, which simply indicated where to place the .component file. In this, it explained _not to highlight a midiO output track_ in GB when using the IAC (loopback) MIDI driver, because having a track highlighted causes that track to accept MIDI input, and it will sometimes accept _any_ MIDI input, including what it's outputting. That creates a nifty feedback loop. The solution is to create an additional blank software instrument track and click on that before attempting any playback.

I'll be pushing my Ruby version of Moppy's controller software (it still uses the same Arduino code) to a new branch on my own fork after I clean it up to something I can be proud of, acquire my fourth drive & finish my first song. Thanks for your help!
